test/test_images/recordevents/eventstore.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	cloudevents "github.com/cloudevents/sdk-go"
+
+	"knative.dev/eventing/test/lib"
+)
+
+// Number of EventInfo per block
+const evBlockSize = 100
+
+// Block of stored EventInfo
+type eventBlock struct {
+	// seqno of [0] evInfoBytes entry
+	firstIndex int
+	// offset inside block for newly appended event
+	firstOffsetFree int
+	// offset inside block of first non-trimmed event
+	firstValid int
+	// serialized EventInfo structures for each seqno.  We enforce
+	// that there is always at least one block.
+	evInfoBytes [evBlockSize][]byte
+}
+
+// All events currently seen
+type eventStore struct {
+	// Blocks of events in increasing sequency number order
+	evBlocks     []*eventBlock
+	evBlocksLock sync.Mutex
+}
+
+// Create a new event store.
+func newEventStore() *eventStore {
+	es := &eventStore{}
+	es.evBlocks = []*eventBlock{&eventBlock{}}
+
+	// One block with no entries starting at sequence number 1
+	es.evBlocks[0].firstIndex = 1
+	es.evBlocks[0].firstOffsetFree = 0
+	es.evBlocks[0].firstValid = 0
+	return es
+}
+
+// See if there's enough room to append to the current last block.  If not,
+// append an extra block.
+func (es *eventStore) checkAppendBlock() {
+	if es.evBlocks[len(es.evBlocks)-1].firstOffsetFree == evBlockSize {
+		newEVBlock := &eventBlock{
+			firstIndex: es.evBlocks[len(es.evBlocks)-1].firstIndex + evBlockSize,
+		}
+		es.evBlocks = append(es.evBlocks, newEVBlock)
+	}
+}
+
+// Store the specified event.
+func (es *eventStore) StoreEvent(event cloudevents.Event, httpHeaders map[string][]string) {
+	var evInfo lib.EventInfo
+	evInfo.Event = &event
+	evInfo.HTTPHeaders = httpHeaders
+	evInfoBytes, err := json.Marshal(&evInfo)
+	if err != nil {
+		evInfo.Event = nil
+		evInfo.ValidationError = err.Error()
+		if evInfo.ValidationError == "" {
+			evInfo.ValidationError = "Unknown Error"
+		}
+		evInfoBytes, err = json.Marshal(&evInfo)
+		if err != nil {
+			panic(fmt.Errorf("Unexpected marshal error (%s) (%+v)", err, evInfo))
+		}
+	}
+
+	es.evBlocksLock.Lock()
+
+	evBlock := es.evBlocks[len(es.evBlocks)-1]
+	if evBlock.firstOffsetFree < evBlockSize {
+		evBlock.evInfoBytes[evBlock.firstOffsetFree] = evInfoBytes
+		evBlock.firstOffsetFree++
+	}
+
+	// We always keep at least one free entry
+	es.checkAppendBlock()
```

Comment:
Hmm, what's the reason for always reserving at least one free entry? Cannot we simply do append?

Reply:
No.  Vestigial code.  shifted it above the store now.  The only thing i really need is to make sure there is always one block to avoid lots of special cases, but the Trim/initialization functions already ensure that.